sql/expression/function/strcmp_test.go

```diff
@@ -36,10 +37,13 @@ func TestStrCmp(t *testing.T) {
 		{"equal strings", sql.Text, sql.Text, sql.NewRow("a", "a"), int(0), nil},
 		{"first string is smaller", sql.Text, sql.Text, sql.NewRow("a", "b"), int(-1), nil},
 		{"second string is smaller", sql.Text, sql.Text, sql.NewRow("b", "a"), int(1), nil},
-		{"arguments have different types", sql.Int8, sql.Text, sql.NewRow(1, "1"), int(0), nil},
 		{"first argument is null", sql.Text, sql.Text, sql.NewRow(nil, "a"), nil, nil},
 		{"second argument is null", sql.Text, sql.Text, sql.NewRow("a", nil), nil, nil},
 		{"both arguments are null", sql.Text, sql.Text, sql.NewRow(nil, nil), nil, nil},
+		{"different collations throws error", sql.MustCreateString(sqltypes.Text, 10, 255), sql.MustCreateString(sqltypes.Text, 10, 309), sql.NewRow("a", "a"), nil, sql.ErrCollationIllegalMix},
+		{"first argument is text, second argument is not text", sql.Text, sql.Date, sql.NewRow("a", 2022-11-04), int(1), nil},
+		{"first argument is not text, second argument is text", sql.Int8, sql.Text, sql.NewRow(1, "1"), int(0), nil},
+		{"both arguments are non-text", sql.Int8, sql.Date, sql.NewRow(3, 2017-04-27), int(1), nil},
```

Comment:
We need the following additional tests:
* Same character set, same case-sensitivity (one for sensitive, another for insensitive)
* Different character sets, same case-sensitivity
* Different character sets, different case-sensitivity

Then we need to add some tests in `/enginetest/queries/charset_collation_engine.go` and `/enginetest/queries/charset_collation_wire.go`. These should focus on handling coercibility correctly. You can reference the docs that I posted earlier for some ideas. All of these results should also be verified in MySQL, so that we're guaranteed that the results expected by the test are exactly the same as the results expected by MySQL. The two files should mostly have the same tests, the `_engine` and `_wire` files just deal with the output a little differently, hence why their results are the same values but with different types. Wire tests also return strings mapped to their appropriate character set.